In [49]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.preprocessing.image import ImageDataGenerator

In [50]:
dataset_path="C:/Users/Sri Chethan/OneDrive/Desktop/ML/CNN/PetImages"

In [51]:
datagen = ImageDataGenerator(
    rescale=1./255,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,

    validation_split=0.2
)

In [52]:
train_data = datagen.flow_from_directory(
    dataset_path,target_size=(128,128),
    batch_size=32,
    class_mode='binary',
    subset='training'
    )

Found 19970 images belonging to 2 classes.


In [53]:
val_data=datagen.flow_from_directory(
    dataset_path,target_size=(128,128),
    batch_size=32,
    class_mode='binary',
    subset='validation'
    )

Found 4991 images belonging to 2 classes.


In [54]:
print(train_data.class_indices)

{'Cat': 0, 'Dog': 1}


In [55]:
model = keras.Sequential()

In [56]:
# first convolution layer
model.add(
    keras.layers.Conv2D(
        filters=32,
        kernel_size=(3,3),
        activation='relu',
        input_shape = (128,128,3)
    )
)
model.add(keras.layers.BatchNormalization())

c:\Users\Sri Chethan\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [57]:
# first pooling layer
model.add(
    keras.layers.MaxPooling2D(pool_size=(2,2))
)

In [58]:
# 2 convolution layer
model.add(
    keras.layers.Conv2D(
        filters=64,
        kernel_size=(3,3),
        activation='relu',
    )
)
model.add(keras.layers.BatchNormalization())

In [59]:
# second pooling layer
model.add(
    keras.layers.MaxPooling2D(
        pool_size=(2,2)
    )
)

In [60]:
model.add(
    keras.layers.Conv2D(
        filters=128,
        kernel_size=(3,3),
        activation='relu',
    )
)
model.add(keras.layers.BatchNormalization())

In [61]:
#  pooling layer
model.add(
    keras.layers.MaxPooling2D(
        pool_size=(2,2)
    )
)

In [62]:
# 4th convolution layer
model.add(
    keras.layers.Conv2D(
        filters=128,
        kernel_size=(3,3),
        activation='relu',
    )
)
model.add(keras.layers.BatchNormalization())

In [63]:
# 4th pooling layer
model.add(
    keras.layers.MaxPooling2D(
        pool_size=(2,2)
    )
)

In [64]:
# 5th convolution layer
model.add(
    keras.layers.Conv2D(
        filters=256,
        kernel_size=(3,3),
        activation='relu',
    )
)
model.add(keras.layers.BatchNormalization())

In [65]:
# 5th pooling layer
model.add(
    keras.layers.MaxPooling2D(
        pool_size=(2,2)
    )
)

In [66]:
# Dropout for regularization
model.add(
    keras.layers.Dropout(0.5)
)

In [67]:
model.add(
    keras.layers.Flatten()
)
model.add(
    keras.layers.Dense(
        units=256,
        activation='relu'
    )
)
model.add(keras.layers.Dropout(0.5))
model.add(
    keras.layers.Dense(
        units=128,
        activation='relu'
    )
)
model.add(keras.layers.Dropout(0.3))

In [68]:
model.add(
    keras.layers.Dense(
        units=1,
        activation='sigmoid'
    )
)

In [69]:
model.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_9 (Conv2D)               │ (None, 126, 126, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_6           │ (None, 126, 126, 32)   │           128 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_8 (MaxPooling2D)  │ (None, 63, 63, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_10 (Conv2D)              │ (None, 61, 61, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_7           │ (None, 61, 61, 64)     │           256 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_9 (MaxPooling2D)  │ (None, 30, 30, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_11 (Conv2D)              │ (None, 28, 28, 128)    │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_8           │ (None, 28, 28, 128)    │           512 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_10 (MaxPooling2D) │ (None, 14, 14, 128)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_12 (Conv2D)              │ (None, 12, 12, 128)    │       147,584 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_9           │ (None, 12, 12, 128)    │           512 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_11 (MaxPooling2D) │ (None, 6, 6, 128)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_13 (Conv2D)              │ (None, 4, 4, 256)      │       295,168 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_10          │ (None, 4, 4, 256)      │         1,024 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_12 (MaxPooling2D) │ (None, 2, 2, 256)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_3 (Dropout)             │ (None, 2, 2, 256)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_2 (Flatten)             │ (None, 1024)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 256)            │       262,400 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_4 (Dropout)             │ (None, 256)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_6 (Dense)                 │ (None, 128)            │        32,896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_5 (Dropout)             │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 1)              │           129 │
└─────────────────────────────────┴────────────────────────┴─────────────

 Total params: 833,857 (3.18 MB)

 Trainable params: 832,641 (3.18 MB)

 Non-trainable params: 1,216 (4.75 KB)

#### (None,126,126,channels)
#### (Batch size,Height,width , no.of filters )

#### first convolution layer
Input = 128*128*3
filter = 3*3

no.of filters = 32
output size formula :

output = ((input-kernel)/(stride))+1

stride=1

output =((128-3)/1)+1
       =126

channels=32

final output = (None,126,126,32)

kernel

formula = (kh*kw*inputchannels+1)*filters
kernel = 3*3
InputChannels:3
Filters : 32
params : (3*3*3+1)*32
       (28*32)=896

#### pooling2D

MaxPooling2D((2,2))

Formula : 126/2=63
 
final : (None,63,63,32)

#### second conv layer

Input from pooling: (None,63*63*32)

kernel : 3*3

output : 63-3 +1
       : 61

(None,61,61,32)

params =(3*3*32+1)*32=9248

#### second pooling layer

(None,30,30,32)
params=0

pooling : 2*2

floor divison : 61/2=30

final : (none,30,30,32)

No Paramters

#### Flatten

input : 30*30*32

output : 28800

#### Dense Layer

input=28800

neurons=128

params=(input*neurons)+bias
       =3686400+128
       =3686528

#### output layer

input =128

output =1

output shape (None,1)
Meaning = Single prediction
-----> Cat or Dog

Parameter Calculation : 
       = 128*1+1=129


Input --> Conv1 --> Pool1 --> Conv2 --> pool2 --> flatten --> dense --> output



In [70]:
print(tf.config.list_physical_devices())

[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')]


In [72]:
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

history = model.fit(
    train_data,
    epochs=5,
    validation_data=val_data
)

loss,accuracy=model.evaluate(val_data)
print("Accuracy :",accuracy)

Epoch 1/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 584s 928ms/step - accuracy: 0.7261 - loss: 0.5481 - val_accuracy: 0.6686 - val_loss: 0.5976
Epoch 2/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 264s 423ms/step - accuracy: 0.8109 - loss: 0.4169 - val_accuracy: 0.7508 - val_loss: 0.5391
Epoch 3/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 232s 371ms/step - accuracy: 0.8616 - loss: 0.3269 - val_accuracy: 0.8605 - val_loss: 0.3319
Epoch 4/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 227s 363ms/step - accuracy: 0.8834 - loss: 0.2785 - val_accuracy: 0.8345 - val_loss: 0.3518
Epoch 5/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 224s 358ms/step - accuracy: 0.8982 - loss: 0.2442 - val_accuracy: 0.7013 - val_loss: 0.5158
156/156 ━━━━━━━━━━━━━━━━━━━━ 20s 130ms/step - accuracy: 0.6983 - loss: 0.5167
Accuracy : 0.6982568502426147
